packages/ra-ui-materialui/src/list/datagrid/DatagridRow.tsx

```diff
@@ -123,42 +114,35 @@ const DatagridRow: React.ForwardRefExoticComponent<
         },
         [id, onToggleItem, selectable]
     );
+
+    const linkType =
+        rowClick === 'expand' || rowClick === 'toggleSelection'
+            ? undefined
+            : typeof rowClick === 'function'
+              ? // rowClick doesn't have the same signature as linkTo, so we need to adapt
+                (record, resource) => rowClick(record?.id, resource, record)
+              : rowClick;
+    const path = useGetPathForRecord({ record, resource, link: linkType });
+
     const handleClick = useCallback(
         async event => {
             event.persist();
-            const type =
-                typeof rowClick === 'function'
-                    ? await rowClick(id, resource, record)
-                    : rowClick;
-            if (type === false || type == null) {
-                return;
-            }
-            if (['edit', 'show'].includes(type)) {
-                navigate(createPath({ resource, id, type }), {
-                    state: { _scrollToTop: true },
-                });
-                return;
-            }
-            if (type === 'expand') {
+            if (rowClick === 'expand') {
                 handleToggleExpand(event);
                 return;
             }
-            if (type === 'toggleSelection') {
+            if (rowClick === 'toggleSelection') {
                 handleToggleSelection(event);
                 return;
             }
-            navigate(type);
+            if (path === false || path == null) {
+                return;
+            }
+            navigate(path, {
+                state: { _scrollToTop: true },
+            });
```

Comment:
This is a breaking change: if the user passed a `path` or a function that returns a `path`, it will have `scrollToTop: true` while it used not to.